types/src/v0/impls/reward.rs

```diff
@@ -0,0 +1,425 @@
+                    RewardMerkleTree::verify(comm.digest(), RewardAccount(self.account), proof)?
+                        .is_ok(),
+                    "invalid proof"
+                );
+                Ok(proof
+                    .elem()
+                    .context("presence proof is missing account balance")?
+                    .0)
+            },
+            RewardMerkleProof::Absence(proof) => {
+                let tree = RewardMerkleTree::from_commitment(comm);
+                ensure!(
+                    tree.non_membership_verify(RewardAccount(self.account), proof)?,
+                    "invalid proof"
+                );
+                Ok(0.into())
+            },
+        }
+    }
+
+    pub fn remember(&self, tree: &mut RewardMerkleTree) -> anyhow::Result<()> {
+        match &self.proof {
+            RewardMerkleProof::Presence(proof) => {
+                tree.remember(
+                    RewardAccount(self.account),
+                    proof
+                        .elem()
+                        .context("presence proof is missing account balance")?,
+                    proof,
+                )?;
+                Ok(())
+            },
+            RewardMerkleProof::Absence(proof) => {
+                tree.non_membership_remember(RewardAccount(self.account), proof)?;
+                Ok(())
+            },
+        }
+    }
+}
+
+impl From<(RewardAccountProof, U256)> for RewardAccountQueryData {
+    fn from((proof, balance): (RewardAccountProof, U256)) -> Self {
+        Self { balance, proof }
+    }
+}
+
+pub fn apply_rewards(
+    mut reward_state: RewardMerkleTree,
+    validator: Validator<BLSPubKey>,
+) -> anyhow::Result<RewardMerkleTree> {
+    let mut update_balance = |account: &RewardAccount, amount: RewardAmount| {
+        reward_state = reward_state.persistent_update_with(account, |balance| {
+            let balance = balance.copied();
+            match balance.unwrap_or_default().0.checked_add(amount.0) {
+                Some(updated) => Some(updated.into()),
+                None => {
+                    tracing::warn!("overflowed reward balance for account {}", account);
+                    balance
+                },
+            }
+        })?;
+        Ok::<(), anyhow::Error>(())
+    };
+
+    // Distribute leader reward
+    let leader_reward = RewardAmount::from(
+        (U256::from(validator.commission) * block_reward().0) / U256::from(COMMISSION_BASIS_POINTS),
+    );
+    update_balance(&RewardAccount(validator.account.to_ethers()), leader_reward)?;
+
+    let delegators_reward = block_reward()
+        .checked_sub(&leader_reward)
+        .context("overflow")?;
+    // Distribute delegator rewards
+    let total_stake = validator.stake.to_ethers();
+    for (delegator_address, delegator_stake) in &validator.delegators {
+        let delegator_reward = RewardAmount::from(
+            delegator_stake.to_ethers() * U256::from(delegators_reward) / total_stake,
```

Comment:
Can we rip the reward computation out into a pure function so we can write some unittests for it? It seems correct to me but I think it would be nice is if we had some tests that we don't lose anything due to rounding errors.

Also we don't have overflow checks for the multiplication and in release mode it will overflow, I think it shouldn't happen here either, but if it's going to happen, here it's more likely here than when updating balances.

Reply:
https://github.com/EspressoSystems/espresso-network/pull/2772/commits/69f42ceca394a959a51930b8d5ad821ba6ff694b